### ISSUE TITLE
Error in TypeScript compilation in 3.0.2

### ISSUE THREAD

**Issue by MarvinRudolph** [greensock/GSAP#341]
**TypeScript Version:** 3.6.4

I use Nuxt 2.10.2 and it works with gsap 3.0.1.
After upgrading to 3.0.2 I have a lot of TS errors with various properties:

```ts
78:5 Property 'bottom' of type 'string | undefined' is not assignable to string index type 'string'.
    76 |   interface GSAPUnits {
    77 |     [key: string]: string;
  > 78 |     bottom?: string;
       |     ^
    79 |     fontSize?: string;
    80 |     height?: string;
    81 |     left?: string;


                                                                                                     nuxt:typescript 11:01:35
79:5 Property 'fontSize' of type 'string | undefined' is not assignable to string index type 'string'.
    77 |     [key: string]: string;
    78 |     bottom?: string;
  > 79 |     fontSize?: string;
       |     ^
    80 |     height?: string;
    81 |     left?: string;
    82 |     lineHeight?: string;


                                                                                                     nuxt:typescript 11:01:35
80:5 Property 'height' of type 'string | undefined' is not assignable to string index type 'string'.
    78 |     bottom?: string;
    79 |     fontSize?: string;
  > 80 |     height?: string;
       |     ^
    81 |     left?: string;
    82 |     lineHeight?: string;
    83 |     margin?: string;


                                                                                                     nuxt:typescript 11:01:35
81:5 Property 'left' of type 'string | undefined' is not assignable to string index type 'string'.
    79 |     fontSize?: string;
    80 |     height?: string;
  > 81 |     left?: string;
       |     ^
    82 |     lineHeight?: string;
    83 |     margin?: string;
    84 |     padding?: string;


                                                                                                     nuxt:typescript 11:01:35
82:5 Property 'lineHeight' of type 'string | undefined' is not assignable to string index type 'string'.
    80 |     height?: string;
    81 |     left?: string;
  > 82 |     lineHeight?: string;
       |     ^
    83 |     margin?: string;
    84 |     padding?: string;
    85 |     perspective?: string;


                                                                                                     nuxt:typescript 11:01:35
83:5 Property 'margin' of type 'string | undefined' is not assignable to string index type 'string'.
    81 |     left?: string;
    82 |     lineHeight?: string;
  > 83 |     margin?: string;
       |     ^
    84 |     padding?: string;
    85 |     perspective?: string;
    86 |     right?: string;


                                                                                                     nuxt:typescript 11:01:35
84:5 Property 'padding' of type 'string | undefined' is not assignable to string index type 'string'.
    82 |     lineHeight?: string;
    83 |     margin?: string;
  > 84 |     padding?: string;
       |     ^
    85 |     perspective?: string;
    86 |     right?: string;
    87 |     rotation?: string;


                                                                                                     nuxt:typescript 11:01:35
85:5 Property 'perspective' of type 'string | undefined' is not assignable to string index type 'string'.
    83 |     margin?: string;
    84 |     padding?: string;
  > 85 |     perspective?: string;
       |     ^
    86 |     right?: string;
    87 |     rotation?: string;
    88 |     rotationX?: string;


                                                                                                     nuxt:typescript 11:01:35
86:5 Property 'right' of type 'string | undefined' is not assignable to string index type 'string'.
    84 |     padding?: string;
    85 |     perspective?: string;
  > 86 |     right?: string;
       |     ^
    87 |     rotation?: string;
    88 |     rotationX?: string;
    89 |     rotationY?: string;


                                                                                                     nuxt:typescript 11:01:35
87:5 Property 'rotation' of type 'string | undefined' is not assignable to string index type 'string'.
    85 |     perspective?: string;
    86 |     right?: string;
  > 87 |     rotation?: string;
       |     ^
    88 |     rotationX?: string;
    89 |     rotationY?: string;
    90 |     skewX?: string;


                                                                                                     nuxt:typescript 11:01:35
88:5 Property 'rotationX' of type 'string | undefined' is not assignable to string index type 'string'.
    86 |     right?: string;
    87 |     rotation?: string;
  > 88 |     rotationX?: string;
       |     ^
    89 |     rotationY?: string;
    90 |     skewX?: string;
    91 |     skewY?: string;


                                                                                                     nuxt:typescript 11:01:35
89:5 Property 'rotationY' of type 'string | undefined' is not assignable to string index type 'string'.
    87 |     rotation?: string;
    88 |     rotationX?: string;
  > 89 |     rotationY?: string;
       |     ^
    90 |     skewX?: string;
    91 |     skewY?: string;
    92 |     top?: string;


                                                                                                     nuxt:typescript 11:01:35
90:5 Property 'skewX' of type 'string | undefined' is not assignable to string index type 'string'.
    88 |     rotationX?: string;
    89 |     rotationY?: string;
  > 90 |     skewX?: string;
       |     ^
    91 |     skewY?: string;
    92 |     top?: string;
    93 |     width?: string;


                                                                                                     nuxt:typescript 11:01:35
91:5 Property 'skewY' of type 'string | undefined' is not assignable to string index type 'string'.
    89 |     rotationY?: string;
    90 |     skewX?: string;
  > 91 |     skewY?: string;
       |     ^
    92 |     top?: string;
    93 |     width?: string;
    94 |     x?: string;


                                                                                                     nuxt:typescript 11:01:35
92:5 Property 'top' of type 'string | undefined' is not assignable to string index type 'string'.
    90 |     skewX?: string;
    91 |     skewY?: string;
  > 92 |     top?: string;
       |     ^
    93 |     width?: string;
    94 |     x?: string;
    95 |     y?: string;


                                                                                                     nuxt:typescript 11:01:35
93:5 Property 'width' of type 'string | undefined' is not assignable to string index type 'string'.
    91 |     skewY?: string;
    92 |     top?: string;
  > 93 |     width?: string;
       |     ^
    94 |     x?: string;
    95 |     y?: string;
    96 |     z?: string;


                                                                                                     nuxt:typescript 11:01:35
94:5 Property 'x' of type 'string | undefined' is not assignable to string index type 'string'.
    92 |     top?: string;
    93 |     width?: string;
  > 94 |     x?: string;
       |     ^
    95 |     y?: string;
    96 |     z?: string;
    97 |   }


                                                                                                     nuxt:typescript 11:01:35
95:5 Property 'y' of type 'string | undefined' is not assignable to string index type 'string'.
    93 |     width?: string;
    94 |     x?: string;
  > 95 |     y?: string;
       |     ^
    96 |     z?: string;
    97 |   }
    98 |


                                                                                                     nuxt:typescript 11:01:35
96:5 Property 'z' of type 'string | undefined' is not assignable to string index type 'string'.
    94 |     x?: string;
    95 |     y?: string;
  > 96 |     z?: string;
       |     ^
    97 |   }
    98 |
    99 |   interface StaggerVars extends CallbackVars, utils.DistributeConfig {


                                                                                                nuxt:typescript 11:01:35
24:5 Property 'duration' of type 'number | InertiaDuration | undefined' is not assignable to string index type 'string | number | FunctionBasedValue<number> | FunctionBasedValue<string> | InertiaDuration | InertiaObject'.
    22 |   interface InertiaVars {
    23 |     [key: string]: TweenValue | InertiaObject | InertiaDuration;
  > 24 |     duration?: number | InertiaDuration;
       |     ^
    25 |     linkedProps?: string;
    26 |     resistance?: number;
    27 |   }


                                                                                                nuxt:typescript 11:01:35
25:5 Property 'linkedProps' of type 'string | undefined' is not assignable to string index type 'string | number | FunctionBasedValue<number> | FunctionBasedValue<string> | InertiaDuration | InertiaObject'.
    23 |     [key: string]: TweenValue | InertiaObject | InertiaDuration;
    24 |     duration?: number | InertiaDuration;
  > 25 |     linkedProps?: string;
       |     ^
    26 |     resistance?: number;
    27 |   }
    28 |


                                                                                                nuxt:typescript 11:01:35
26:5 Property 'resistance' of type 'number | undefined' is not assignable to string index type 'string | number | FunctionBasedValue<number> | FunctionBasedValue<string> | InertiaDuration | InertiaObject'.
    24 |     duration?: number | InertiaDuration;
    25 |     linkedProps?: string;
  > 26 |     resistance?: number;
       |     ^
    27 |   }
    28 |
    29 |   interface TweenVars {
```

**Comment by OSUblake** [comment]
Do you have strictNullChecks on in your tsconfig? Try turning that off.

**Comment by OSUblake** [comment]
I can add `undefined` as a property value in the next release, but that's really not the intent of how gsap is supposed to used.

I really don't think `undefined` should be a valid property value, but with `strictNullChecks` it is.

```js
gsap.to(".class", {
  x: undefined
});
```

**Comment by elswenson** [comment]
@OSUblake I appreciate your fast response to this github issue, as I also ran into this issue when upgrading to 3.0.2. I think you are correct that turning off `strictNullChecks` would fix this; however that is not a solution many Typescript users will accept, nor one I think they should accept or need to accept to use gsap.

I took a quick look at the types directory, would this work as a solution? With the way GSAPUnits is being used, it could be turned into a type. 

```typescript
type GSAPUnits = {
  bottom?: string;
  fontSize?: string;
  height?: string;
  left?: string;
  lineHeight?: string;
  margin?: string;
  padding?: string;
  perspective?: string;
  right?: string;
  rotation?: string;
  rotationX?: string;
  rotationY?: string;
  skewX?: string;
  skewY?: string;
  top?: string;
  width?: string;
  x?: string;
  y?: string;
  z?: string;
} & {
  [key: string]: string
}
```

I'm not sure what the intended behavior of the interface in 3.0.2 is, but my proposed type intersection  would behave as follows: 
```typescript
const x = (vars: GSAPUnits) => {
  // Typescript error: May be undefined due to the property being optional
  vars.right.toString();
  // No typescript error, as it's not a key defined above as potentially undefined. 
  vars.NotADefinedProperty.toString();
};
```

**Comment by OSUblake** [comment]
@elswenson

Setting `strictNullChecks` was meant as a temp workaround while we work on getting this fixed. 

I'll look into the fix you posted. The intended behavior for GSAPUnits is like this. You can tell gsap what the default CSS unit should be for a given property. Some properties are already defined, like the ones listed in the interface, but you can set any property. 

```js
gsap.config({
  units: {
    x: "%",
    foo: "px"
  }
});
```

**Comment by MarvinRudolph** [comment]
I think I will just stick to 3.0.1 until it is fixed. `strictNullChecks` would be okay as temp workaround anyways. :)

**Comment by jackdoyle** [comment]
This should be fixed in the latest release (3.0.3)
